cubi_tk/sodar/check_remote.py

```diff
@@ -0,0 +1,441 @@
+"""``cubi-tk sodar check-remote``: check that files are present in remote SODAR/iRODS.
+
+This tool searches recursively for all files with associated .md5 files in a given directory (or PWD)
+and compares them against files linked to a given SODAR project/irods directory, also the md5 sums
+recorded in local .md5 files can be rechecked. No hash format other than md5 is currently supported.
+Comparison is done based (first) on file names and then (unless disabled) on md5 sums,
+therefore wiles with different names but the same md5 sum will not be matched.
+File are sorted into:
+- those present only locally
+- those present both locally and in the remote directory
+- those present only in the remote directory
+"""
+import argparse
+from collections import defaultdict
+import os
+from pathlib import Path
+from types import SimpleNamespace
+import typing
+import attr
+
+from logzero import logger
+
+from ..common import load_toml_config, compute_md5_checksum
+from ..exceptions import FileMd5MismatchException
+from ..snappy.retrieve_irods_collection import RetrieveIrodsCollection
+from ..snappy.check_remote import Checker as SnappyChecker
+
+# RetrieveIrodsCollection needs this in the namespace
+# either define it explicitly or import it
+DEFAULT_HASH_SCHEME = "MD5"
+
+
+# Adapted from snappy.retrieve_irods_collection
+@attr.s(frozen=True, auto_attribs=True)
+class FileDataObject:
+    """File data object - simple container to keep track of files information and allow comparison"""
+
+    file_name: str
+    file_path: str
+    file_md5sum: str
+
+
+class FindLocalMD5Files:
+    """Class contains methods to find local files with associated md5 sums"""
+
+    def __init__(self, base_path, recheck_md5=False):
+        """Constructor: init vars"""
+
+        self.searchpath = Path(base_path)
+        self.recheck_md5 = recheck_md5
+
+    # Adapted from snappy check remote
+    def run(self):
+        """Runs class routines.
+
+        :return: Returns dictionary of dictionaries:
+        key: directory path ; value: list of FileDataObject (one per file in that path)
+        """
+        logger.info("Starting raw data files search ...")
+
+        # Initialise variables
+        rawdata_structure_dict = defaultdict(list)
+
+        # Validate input
+        if not self.searchpath.exists():
```

Comment:
Shouldn't this be together with the other argument validations in `SodarCheckRemoteCommand.check_args()`?

Reply:
`SodarCheckRemoteCommand.check_args()` already does check that this path exists - though I think I actually wrote this part earlier.
I guess it's better to remove it here then? Or should the class check anyway so that it - if imported somewhere else - can log the error (and retrun None). Or would it be better to not prevent that error and let it 'crash'.

Reply:
Probably remove. Ii may never happen, but if someone in the future decides to use let that person deal with it.